### ISSUE TITLE
Intermittent files missing in Azure Storage Blob after upload

### ISSUE THREAD

**Issue by LiamDHall** [LanceMcCarthy/Action-AzureBlobUpload#352]
After excuting this step in my github action.
     
      - uses: LanceMcCarthy/Action-AzureBlobUpload@v2
        name: Upload application files to Azure Blob
        id: upload
        with:
          connection_string: ${{ secrets.AZURE_CONNECTION_STRING }}
          container_name: '${{ env.Software_Publish_Container }}'  
          destination_folder: '${{ env.Software_Publish_Blob }}\Application Files'
          source_folder: '${{ env.Publish_Path }}\Application Files'
          clean_destination_folder: false
          is_recursive: true
          delete_if_exists: false
          
when I have then come to check the  destination_folder in Azure Storage Explorer sometimes files are missing. 

I thought maybe there would be a delay in them appearing but after waiting 10mins and refreshing the files did not appear.
     
These files are dll and needed for the deployment of the application I am uploading
     
The source_folder structure is sort of like the following
     
     - Application Files
           - Application Name_1_0_0_0
               - Applicatoin Folder
               - Applicatoin Folder
               - Applicatoin Folder
               - Applicatoin Folder
               - Applicatoin Folder
               - a.dll
               - b.dll
               - c.dll
               - d.dll
               - e.dll
               - application name.manifest
      

Sometimes it misses a couple dll file or it misses the .manifest file

The print out of the github action says all of them uploaded successfully.

Similar to this issue:

https://github.com/LanceMcCarthy/Action-AzureBlobUpload/issues/113

Please advise how best to fix this.

**Comment by LanceMcCarthy** [comment]
Hi Liam, thank you for taking the time to prepare and share those details. 
The first time this happened, it was because of Azure's policy on snapsopts. However in your case, you are not removing the last build before uploading the new one.

In this case, the problematic files seem to always have the same name: the appx/app manifest and the output DLLs will always have the same file name. so I suspect this is due to not having the ability overwrite what you already have in that destination folder.

As a workaround while I research this, can you please try setting `clean_destination_folder: true`

Since you're doing a build and uploading the same files again anyways, it shouldn't have any effect on the final outcome... you will just be ensuring that all the destination files are always the latest verion.

**Comment by LanceMcCarthy** [comment]
I just had a follow-up thought. It seems some of the affected items have spaces in their files paths.... this might be triggering the auto-fix mechanism that prevents duplicate paths (and stripping out the duplicate).

Can you please also try using an underscore in the file paths. For example, notice how I've replaces spaces with an underscore

```
- Application_Files
       - Application_Name_1_0_0_0
           - Application_Folder
           - Application_Folder
           - Application_Folder
           - Application_Folder
           - Application_Folder
           - a.dll
           - b.dll
           - c.dll
           - d.dll
           - e.dll
           - application_name.manifest
```

If you must keep the spaces, try using a URL-encoded space character so that it is safe for transit

```
- Application%20Files
       - Application%20Name_1_0_0_0
           - Application%20Folder
           - Application%20Folder
           - Application%20Folder
           - Application%20Folder
           - Application%20Folder
           - a.dll
           - b.dll
           - c.dll
           - d.dll
           - e.dll
           - application%20name.manifest
```

I have started investigating and will attempt to replicate your results by using the same target path names.

**Comment by LiamDHall** [comment]
Cheers I will try and implement that now and let you know how I get on.

**Comment by LiamDHall** [comment]
If it help in your investigation, each time we publish, a new folder is created with the corresponding version after the Application Name.

For example:

- Application Files
       - Application Name_1_0_0_0
       - Application Name_1_0_0_1
       - Application Name_1_0_0_2
       - Application Name_1_0_0_3
       - Application Name_1_0_0_4
       and so on
      
each have the foremetioned file structure inside them.

**Comment by LiamDHall** [comment]
When we published into a pre-existing folder say  "Application Name_1_0_0_3" 

with: clean_destination_folder: true

it worked 3 out of 3 times.

once we published and a new folder created say Application Name_1_0_0_5 with the upload it missed files.

this time the file missed was:
![github actions](https://github.com/LanceMcCarthy/Action-AzureBlobUpload/assets/63807594/9a0dcaa6-69a5-4f41-98e1-8c792a97269c)
 zh-Hans/Microsoft.Data.Edm.resources.dll.deploy

**Comment by LanceMcCarthy** [comment]
Aha! Okay so your issue is because you're not delete the duplicate file (e.g., the manifest will still be the old one).

To address this, please set `delete_if_exists` to `true` (instead of `false`).

this will ensure that the newer app.manifest file (and newer build dll) gets uploaded and replaces the old one. Don't worry about the other "Application Name" folders, they will stay there because you have `clean_destination_folder: false`

Here's what the updated task would look like:

```yaml
- uses: LanceMcCarthy/Action-AzureBlobUpload@v2
    name: Upload application files to Azure Blob
    id: upload
    with:
      connection_string: ${{ secrets.AZURE_CONNECTION_STRING }}
      container_name: '${{ env.Software_Publish_Container }}'  
      destination_folder: '${{ env.Software_Publish_Blob }}\Application Files'
      source_folder: '${{ env.Publish_Path }}\Application Files'
      clean_destination_folder: false
      is_recursive: true
      delete_if_exists: true
```

**Comment by LiamDHall** [comment]
Hi 

I believe the following set up has worked as you suggested!

  - uses: LanceMcCarthy/Action-AzureBlobUpload@v2
      name: Upload application files to Azure Blob
      id: upload
      with:
        connection_string: ${{ secrets.AZURE_CONNECTION_STRING }}
        container_name: '${{ env.Software_Publish_Container }}'  
        destination_folder: '${{ env.Software_Publish_Blob }}\Application Files'
        source_folder: '${{ env.Publish_Path }}\Application Files'
        clean_destination_folder: false
        is_recursive: true
        delete_if_exists: true
        
with 3 out 3 successfull publish.

Thank you so much for your help! 👍 

As this is an intermittent, please leave the issue open for a little while and if an upload fails I will let you know :).

For my understanding please could you explain why setting delete_if_exists to true made a differences for files uploaded to folder created by the upload?

In my mind mine it shouldn't matter as no files existed as the folder didn't exist before the upload so nothing would have existed if that makes sence.

so if uploaded version 2.0.0.1 the folder Application Name_2_0_0_1 will be create by the upload.

**Comment by LanceMcCarthy** [comment]
Here's what I think might be happening:

Because you have `clean_destination_folder: false`, all of the files from the last upload are still in the destination blob when you upload again.

As an examples, let's say you have **One.txt** in the blob folder already. Now you've generated a new version of **One.txt** and also a new Two.txt that you want to add. This is what can happen sometimes:

One.txt  <- this one is still old because the Azure SDK did not replace it with the new one
Two.txt <-- this one is new, so it gets uploaded without a problem

However, if you set `delete_if_exists: true`, then it will delete **One.txt** before it uploads the new copy. This always gives you the latest copy of the file with the latest contents.

Note that's just a guess as to what is going on here. This Action has turned out to be a lot more complicated that when I first developed it :D So, I've added more functionality and features over time that handle these edge cases for certain scenarios. Most of the time, it was because of the way the SDK works, but sometimes it's because of the Blob structure/rules.

We can leave this open for a couple weeks, just in case something goes wrong again.
